actionview/lib/action_view/helpers/asset_url_helper.rb

```diff
@@ -118,16 +118,67 @@ module Helpers
-      #   asset_path "http://www.example.com/js/xmlhr.js" # => http://www.example.com/js/xmlhr.js
+      # === With the asset pipeline
+      #
+      # All options passed to +asset_path+ will be passed to +compute_asset_path+
+      # which is implemented by sprockets-rails.
+      #
+      #      asset_path("application.js") # => "/assets/application-60aa4fdc5cea14baf5400fba1abf4f2a46a5166bad4772b1effe341570f07de9.js"
+      #
+      # === Without the asset pipeline (<tt>skip_pipeline: true</tt>)
+      #
+      # Accepts a <tt>type</tt> option that can specify the asset's extension. No error
+      # checking is done to verify the source passed into +asset_path+ is valid
+      # and that the file exists on disk.
+      #
+      #      asset_path("application.js", skip_pipeline: true)                 # => "application.js"
+      #      asset_path("filedoesnotexist.png", skip_pipeline: true)           # => "filedoesnotexist.png"
+      #      asset_path("application", type: :javascript, skip_pipeline: true) # => "/javascripts/application.js"
+      #      asset_path("application", type: :stylesheet, skip_pipeline: true) # => "/stylesheets/application.css"
+      #
+      # === Options applying to all assets
+      #
+      # Below lists scenarios that apply to +asset_path+ whether or not you're
+      # using the asset pipeline.
+      #
+      # - All fully qualified urls are returned immediately. This bypasses the
+      #   asset pipeline and all other behavior described.
+      #
+      #     asset_path("http://www.example.com/js/xmlhr.js") # => "http://www.example.com/js/xmlhr.js"
+      #
+      # - All assets that begin with a forward slash are assumed to be full
+      #   urls and will not be expanded. This will bypass the asset pipeline.
+      #
+      #     asset_path("/foo.png") # => "/foo.png"
+      #
+      # - All blank strings will be returned immediately. This bypasses the
+      #   asset pipeline and all other behavior described.
+      #
+      #     asset_path("") # => ""
+      #
+      # - If <tt>config.relative_url_root</tt> is specified, all assets will have that
+      #   root prepended.
+      #
+      #     Rails.application.config.relative_url_root = "bar"
+      #     asset_path("foo.js", skip_pipeline: true) # => "bar/foo.js"
+      #
+      # - A different asset host can be specified via <tt>config.action_controller.asset_host</tt>
+      #   this is commonly used in conjunction with a CDN.
+      #
+      #     Rails.application.config.action_controller.asset_host = "assets.example.com"
+      #     asset_path("foo.js", skip_pipeline: true) # => "http://assets.example.com/foo.js"
+      #
+      # - An extension name can be specified manually with <tt>extname</tt>.
+      #
+      #     asset_path("foo", skip_pipeline: true, extname: ".js")     # => "/foo.js"
+      #     asset_path("foo.css", skip_pipeline: true, extname: ".js") # => "/foo.css.js"
```

Comment:
The assetpipeline does a bunch of things and I don't want asset pipeline magic being confused with this option. Adding a long fingerprint to the example makes it less clear that an extension was added to the output. What it doesn't do is replace an extension name, so `foo.css` does not become `foo.js`. Also technically this option doesn't change the extension output, it changes the input that goes into the asset pipeline. It's harder to show.